Added support for `-force-color` to the CLI.

apollo13 opened the pull request:
This should fix #10973

apollo13 commented:
@jrasell Would appreciate a review. I have verified manually and it works, not to sure about tests. On the other hand I couldn't really find any for `-no-color` either :D

apollo13 commented:
I am pretty sure that the CSI snapshot failures in the test are not my fault! ;)

greut commented on website/content/partials/general_options.mdx:
```diff
@@ -13,6 +13,10 @@
   `NOMAD_CLI_NO_COLOR` may be set.
 
+-`-force-color`: Forces colored command output. This can be used in cases where
+  the usual	terminal detection fails (CI systems mainly).	Alternatively,
```
```suggestion
  the usual terminal detection fails (CI systems mainly).	Alternatively,
```

greut commented on website/content/partials/general_options_no_namespace.mdx:
```diff
@@ -8,6 +8,10 @@
   `NOMAD_CLI_NO_COLOR` may be set.
 
+-`-force-color`: Forces colored command output. This can be used in cases where
+  the usual	terminal detection fails (CI systems mainly).	Alternatively,
```
```suggestion
  the usual terminal detection fails (CI systems mainly).	Alternatively,
```

apollo13 replied:
Thanks that paragraph had another 3 issues like that as well, not sure how/why that hapenned :D

apollo13 commented:
I does not yet seem to work in gitlab CI, mhm :/ That will be fun to figure out.

apollo13 commented:
Okay, so `-force-color` does work, but the env variable does not. This seems to be caused by the fact that `plan` calls `Meta.Colorize()` which only takes flags into account but not the matching env variables. I am not 100% sure how nomad maps env vars to flags, any pointers?

apollo13 commented:
> Okay, so `-force-color` does work, but the env variable does not. This seems to be caused by the fact that `plan` calls `Meta.Colorize()` which only takes flags into account but not the matching env variables. I am not 100% sure how nomad maps env vars to flags, any pointers?

Fixed for this PR, should be ready for review now

jrasell commented on command/meta.go:
```diff
@@ -63,13 +66,19 @@ type Meta struct {
 func (m *Meta) FlagSet(n string, fs FlagSetFlags) *flag.FlagSet {
 	f := flag.NewFlagSet(n, flag.ContinueOnError)
 
+	forceColor := false
+	if os.Getenv(EnvNomadCLIForceColor) != "" {
+		forceColor = true
+	}
```
This seems like it is being used to set the default value on the flag, whereas the default value should be fixed and not mutable by env vars. I think we can remove this line and update the default flag value to be false.

apollo13 replied:
Simply changing `meta.forceColor` to `*bool` and fixing up the code so it compiles crashes at runtime in `FlagSet`:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1f08e75]

goroutine 1 [running]:
flag.newBoolValue(...)
	flag/flag.go:112
flag.(*FlagSet).BoolVar(...)
	flag/flag.go:630
github.com/hashicorp/nomad/command.(*Meta).FlagSet(0xc00078da00, 0x2862ddc, 0x8, 0x2, 0x4c0b58)
	github.com/hashicorp/nomad/command/meta.go:75 +0x1d5
github.com/hashicorp/nomad/command.(*JobPlanCommand).Run(0xc00078da00, 0xc0001b2030, 0x1, 0x1, 0xc0003465c0)
	github.com/hashicorp/nomad/command/job_plan.go:122 +0xf1
github.com/mitchellh/cli.(*CLI).Run(0xc00034ea00, 0xc00034ea00, 0xc0001a0300, 0x43)
	github.com/mitchellh/cli@v1.1.0/cli.go:260 +0x41a
main.RunCustom(0xc0001b2010, 0x3, 0x3, 0xc000180058)
	github.com/hashicorp/nomad/main.go:147 +0x55c
main.Run(...)
	github.com/hashicorp/nomad/main.go:87
main.main()
	github.com/hashicorp/nomad/main.go:83 +0x65
```

Do I need to do something else?

apollo13 replied:
Also I do not think using a pointer would work:
```
package main

import (
	"flag"
	"fmt"
)

func main() {
	testFlag := true
	flag.BoolVar(&testFlag, "flagname", false, "help message for flagname")
	fmt.Println(testFlag)
}
```
this would set `testFlag` to false, independent on whether passed or not and I cannot simply pass `nil` as default either.

jrasell replied:
I think it would need a custom flag helper; but having spent a little more time thinking about it I don't think we need to be able to tell the difference between a user setting false, and the default value. Lets therefore ignore the idea of using a pointer and focus on not mutating the default value and using the env var at a later stage.

apollo13 replied:
Ok, I'll be on holiday next week but after that I will see what I can do. How "expensive" are calls to the system env in go? `Colorize` is used quite often, even during plan and while it may not immediately be noticable I rather not have plan call the getenv system call 100 times to render the output.

jrasell reviewed (changes requested):
Thanks a lot for this @apollo13 and sorry for the delay in reviewing. I have an line comment which I would appreciate your thoughts on.

Another comment which appears a couple of times would be to remove the `(CI systems mainly)` statement in the help outputs. I'd prefer the user to understand why they need to use the flag, rather than read the help and maybe assume they can override because of the CI detail. Again I am open to opinions here.

apollo13 commented:
I have fixed up the flag defaults and it should work as expected now. If there are performance issues we have to reevaluate.

apollo13 commented:
@lgfa29 I have rebased this PR onto the latest no-color changes in main. While on it I have simplified the code because `Colorize` doesn't need to check for a terminal since `meta.Ui` is only set when a terminal exists: https://github.com/hashicorp/nomad/blob/main/main.go#L118 (this is checked in main.go). Given that main.go already parses `-no-color` itself and puts it into the env there is also no reason to check the `no-color` flag (which now makes the test fail due to the way they are written :/).

I like how my PR looks now because it only bothers about the flags and env variables in one place. Sadly this "one place" is main.go which probably makes this somewhat annoying to test. I do not really have an idea on how to fix this -- I am open for ideas :) I guess we could somewhat "fix" this by moving the creation of `meta.Ui` into an utility function that could be tested more easily.

apollo13 commented:
I have pushed a new commit with a `setupUi` utility function, cleaned up the code and fixed the tests. Let me know what you think.

apollo13 commented:
@lgfa29 Any chance you could review this PR? :)

lgfa29 commented:
> @lgfa29 Any chance you could review this PR? :)

Will do it in the coming days 🙂 
Sorry of the delay here.

lgfa29 reviewed (approved):
Thank you for this great PR @apollo13! 

It's great to see this logic being pulled out of `main.go` since it's now easier to test it 🙂 

I pushed a commit to change the logic slightly so that `-no-color` takes precedence of `-force-color`. These two options are meant to be used together, but in case they are it would be better to avoid colors since it's a safer option. 

What do you think?

apollo13 commented:
Personally I think that `force` is stronger than `no` so I would have expected `force` to win, but no strong feelings…

lgfa29 commented:
> Personally I think that `force` is stronger than `no` so I would have expected `force` to win, but no strong feelings…

So now we need a `-force-no-color` to beat them all 😅 

Just kidding, or course. The reasoning behind `-no-color` to win is that using the two flags at the same time is most likely a mistake, and `-no-color` is the safest option since it won't pollute the output in case colors are not supported.

I will merge this now then. Thank you for the contribution!
